awscli/customizations/logs/tail.py

```diff
@@ -294,20 +295,31 @@ def _filter_log_events(self, filter_logs_events_kwargs):
             return
 
     def _do_filter_log_events(self, filter_logs_events_kwargs):
-        last_event_ids = []
+        event_ids_per_timestamp = defaultdict(set)
         while True:
             response = self._client.filter_log_events(
                 **filter_logs_events_kwargs)
             for event in response['events']:
                 # For the case where we've hit the last page, we will be
-                # reusing the most recent nextToken to continue polling. This
-                # means it is possible that duplicate log events from the last
-                # page are returned back which we do not want to yield again.
+                # reusing the newest timestamp of the received events to keep polling.
+                # This means it is possible that duplicate log events with same timestamp
+                # are returned back which we do not want to yield again.
                 # We only want to yield log events that we have not seen.
-                if event['eventId'] not in last_event_ids:
+                if event['eventId'] not in event_ids_per_timestamp[event['timestamp']]:
+                    event_ids_per_timestamp[event['timestamp']].add(event['eventId'])
                     yield event
-            last_event_ids = [event['eventId'] for event in response['events']]
+            if event_ids_per_timestamp.keys():
+                # Store ids of the events with the newest timestamp only
+                newest_timestamp = max(event_ids_per_timestamp.keys())
+                event_ids_per_timestamp = defaultdict(
+                    set, {newest_timestamp: event_ids_per_timestamp[newest_timestamp]}
+                )
             if 'nextToken' in response:
                 filter_logs_events_kwargs['nextToken'] = response['nextToken']
             else:
+                # Remove nextToken and update startTime for the next request
```

Comment:
I'm not sure if this feedback will apply if we decide to update the startTime logic, but the `_do_filter_log_events` is getting a little difficult to read. It may be worth starting to break down the more complicated logic into smaller internal methods. For example with this code, I can see there being a:

* `_reset_filter_log_events_params()` which would do the logic of adding a new startTime and popping off the next token.
* `_update_latest_events_and_timestamp()` which would do the book keeping logic that keeps track of what the latest timestamp and events are.

There may be more methods we can come up with, but a good indication if it may be worth adding one is if you find yourself needing to add a comment on why a line of code is needed as often times the method name can do the documenting for us.